### DISCUSSION TITLE
Firebase storage offline support

### DISCUSSION THREAD

**Discussion by meliodev** [invertase/react-native-firebase#4835] [Q&A]
Hi guys !
I need to handle the case where the user tries to upload a file in offline mode. It seems that Firebase Storage doesn't queue the upload task on local cache and persist it once back online (as Firestore does). Is there a workaround to handle that ?

**Comment by mikehardy** [comment]
Interesting.

"It seems" -> exactly what tests have you run that show expectation vs result? Do you have an App.js that demonstrates the difference between expectation and reality in a self-contained way? That would help make things clear

Is this actually supported in the underlying SDKs? Specifically I mean *starting* an upload while network doesn't exist? Most of the docs I read say things will *continue* across network issues or even activity restarts (e.g. https://firebase.google.com/docs/storage/android/upload-files#handle_activity_lifecycle_changes - though I'm not sure we reattach listeners correctly here...) and *maybe* (if we implemented it here, unsure) across process restarts.

But I don't see an indication of whether the SDK supports *starting* an upload while offline. You can check that with a quick throwaway off their quickstart https://github.com/firebase/quickstart-android/tree/master/storage (there are similar quickstarts for ios)

So - a couple things, we need to see exactly what functionality you're trying to make work, and we must know if the underlying SDKs support it before it is actionable here. If you can provide those things, we can move forward I think

**Reply by meliodev** [reply]
Thanks for your support @mikehardy ! I just figured out that the issue occurs only when app is killed before the upload task finishes (No matter if the device is online or offline). The upload task doesn't run when opening the app after it was killed.
Is there a way to resume/start the upload task once app is opened after it was killed ?

**Reply by mikehardy** [reply]
Interesting! There is some support in the APIs I *think* for checking to see if there were any pending storage tasks in order to resume them but I'm not sure we do anything like that / handle it here. Pretty sure we don't actually

**Reply by meliodev** [reply]
So I guess I have to use something like redux-saga to handle this use case ?

**Reply by mikehardy** [reply]
if there is *no* support in firebase-android-sdk / firebase-ios-sdk in the form of APIs that let you query "the native firebase SDK storage module" for interrupted uploads, then yes, you would have to store state somewhere in your app, and manage that state. 

If there *is* support in the underlying firebase SDKs for this (there might be? I did not look super-close at the APIs linked above) then it might be easier to implement the real feature here as a PR to expose those underlying SDK APIs as react-native-firebase APIs, then instead of redux-saga you would have some logic on app boot that said, "hey `@react-native-firebase/storage` were there uploads that did not complete?" and do it that way

**Reply by meliodev** [reply]
Thanks! I will try and get back with my results.

**Comment by NomadicDeveloper22** [comment]


**Reply by lucianolopezz** [reply]
Can you show us an example?
